src/v/iceberg/compatibility_utils.h

```diff
@@ -0,0 +1,170 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "base/vassert.h"
+#include "iceberg/compatibility_types.h"
+#include "iceberg/datatypes.h"
+#include "iceberg/field_collecting_visitor.h"
+
+#include <seastar/util/variant_utils.hh>
+
+namespace iceberg {
+
+namespace detail {
+
+template<typename Fn, typename T>
+concept VoidFieldFn = requires(T t, Fn fn) {
+    { fn(t) } -> std::same_as<void>;
+};
+
+template<typename Fn, typename T>
+concept CheckedFieldFn = requires(T t, Fn fn) {
+    { fn(t) } -> std::same_as<schema_errc_result>;
+};
+
+template<typename Fn>
+concept FieldPredicate = requires(const nested_field* t, Fn fn) {
+    { fn(t) } -> std::same_as<bool>;
+};
+
+template<typename Fn, typename T>
+concept FieldOp = VoidFieldFn<Fn, T> || CheckedFieldFn<Fn, T>;
+
+template<typename T>
+concept StructType = std::is_same_v<std::decay_t<T>, struct_type>;
+
+template<typename T>
+concept NestedFieldType = std::is_same_v<std::decay_t<T>, nested_field>;
+
+template<NestedFieldType T>
+requires std::is_same_v<std::decay_t<T>, nested_field>
+schema_errc_result for_each_field_impl(
+  T* field,
+  detail::FieldOp<T*> auto&& fn,
+  detail::FieldPredicate auto&& filter,
+  chunked_vector<T*>& stk) {
+    constexpr bool is_void_fn = detail::VoidFieldFn<decltype(fn), T*>;
+
+    stk.emplace_back(field);
+    while (!stk.empty()) {
+        auto* dst = stk.back();
+        stk.pop_back();
+        if (dst == nullptr) {
+            return schema_evolution_errc::null_nested_field;
+        } else if (!std::invoke(filter, dst)) {
+            continue;
+        }
+        if constexpr (is_void_fn) {
+            std::invoke(fn, dst);
+        } else if (auto res = std::invoke(fn, dst); res.has_error()) {
+            return res.error();
+        }
+
+        if constexpr (std::is_const_v<T>) {
+            std::visit(reverse_const_field_collecting_visitor{stk}, dst->type);
+        } else {
+            std::visit(reverse_field_collecting_visitor(stk), dst->type);
+        }
+    }
+    return std::nullopt;
+}
+
+template<NestedFieldType T, StructType S>
+schema_errc_result for_each_field_impl(
+  S& s, detail::FieldOp<T*> auto&& fn, detail::FieldPredicate auto&& filter) {
+    static_assert(std::is_const_v<S> == std::is_const_v<T>);
+    chunked_vector<T*> stk;
+    for (const auto& f : std::ranges::reverse_view(s.fields)) {
```

Comment:
Interesting that this was reversed between revisions without too much recourse. Curious what the implication is/was?

Reply:
Nothing functional. Just an accidental omission - I intended to duplicate the control flow of the field collector visitor, though it shouldn't have any effect here.